File: backend/schemas/tech_navigator.py
"""Pydantic schemas for the Tech Navigator scoring API [A-TN-04].

The PUT body is fully partial — any subset of fields can be updated; omitted
fields preserve their existing value. The GET response carries both the raw
sub-criteria entered by the user and the recomputed composites, alongside a
snapshot of the active admin weights so the UI can render attribution.
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


# Reusable type aliases. Sub-criteria are integers 1-5 per spec; rubric labels
# at intermediate values are defined in the KB Tech Navigator slide deck.
SubCriterion = int  # validated via Field(ge=1, le=5) in the model
TransformationLevel = Literal["T0", "T1", "T2"]
ProjectType = Literal[1, 2, 3]


class TechNavigatorUpdate(BaseModel):
    """Partial-update request body for PUT /api/projects/{id}/tech-navigator.

    All fields are optional; only fields present in the request are updated.
    """

    project_type: Optional[ProjectType] = None
    transformation_level: Optional[TransformationLevel] = None
    tn_standardization: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_usage: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_maintenance: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_financial_benefit: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_payback: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_competitive_advantage: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_value_reserved_1: Optional[SubCriterion] = Field(default=None, ge=1, le=5)
    tn_value_reserved_2: Optional[SubCriterion] = Field(default=None, ge=1, le=5)


class ComplexityWeights(BaseModel):
    standardization: float
    usage: float
    maintenance: float


class ValueCreationWeights(BaseModel):
    financial: float
    payback: float
    competitive: float


class RankingWeights(BaseModel):
    value: float
    complexity: float


class TshirtThresholds(BaseModel):
    xs_max: int
    s_max: int
    m_max: int
    l_max: int


class WeightsSnapshotResponse(BaseModel):
    """Snapshot of active admin weights, embedded in the read response."""

    complexity: ComplexityWeights
    value_creation: ValueCreationWeights
    ranking: RankingWeights
    tshirt: TshirtThresholds


class TechNavigatorResponse(BaseModel):
    """Full Tech Navigator profile for a project, including computed scores."""

    project_id: str
    project_type: Optional[ProjectType]
    transformation_level: Optional[TransformationLevel]

    # Raw sub-criteria
    tn_standardization: Optional[int]
    tn_usage: Optional[int]
    tn_maintenance: Optional[int]
    tn_financial_benefit: Optional[int]
    tn_payback: Optional[int]
    tn_competitive_advantage: Optional[int]
    tn_value_reserved_1: Optional[int]
    tn_value_reserved_2: Optional[int]

    # Denormalized computed scores
    complexity_score: Optional[float]
    value_creation_score: Optional[float]
    composite_score: Optional[float]

    # Derived from total_budget
    tshirt_size: Optional[Literal["XS", "S", "M", "L", "XL"]]
    total_budget: Optional[float]

    # Active weights at read time
    weights: WeightsSnapshotResponse

File: frontend/src/types/techNavigator.ts
/**
 * Tech Navigator types — frontend mirror of `backend/schemas/tech_navigator.py`.
 *
 * Spec references: [A-TN-01]..[A-TN-09].
 *
 * The five sub-criteria values and the two reserved slots are all integers 1-5
 * (or null while unscored). Computed scores are floats (rounded to 2 decimals
 * server-side). The composite_score is null until BOTH axis scores are non-null
 * — partial profiles do not contribute to ranking [A-PRI-01].
 */

export type TransformationLevel = 'T0' | 'T1' | 'T2';
export type ProjectType = 1 | 2 | 3;
export type TshirtSize = 'XS' | 'S' | 'M' | 'L' | 'XL';

/** Integer 1-5 sub-criterion score, or null while unscored. */
export type SubCriterionScore = 1 | 2 | 3 | 4 | 5 | null;

export interface ComplexityWeights {
  standardization: number;
  usage: number;
  maintenance: number;
}

export interface ValueCreationWeights {
  financial: number;
  payback: number;
  competitive: number;
}

export interface RankingWeights {
  value: number;
  complexity: number;
}

export interface TshirtThresholds {
  xs_max: number;
  s_max: number;
  m_max: number;
  l_max: number;
}

export interface TechNavigatorWeights {
  complexity: ComplexityWeights;
  value_creation: ValueCreationWeights;
  ranking: RankingWeights;
  tshirt: TshirtThresholds;
}

export interface TechNavigatorProfile {
  project_id: string;
  project_type: ProjectType | null;
  transformation_level: TransformationLevel | null;

  // Raw sub-criteria (1-5 or null)
  tn_standardization: SubCriterionScore;
  tn_usage: SubCriterionScore;
  tn_maintenance: SubCriterionScore;
  tn_financial_benefit: SubCriterionScore;
  tn_payback: SubCriterionScore;
  tn_competitive_advantage: SubCriterionScore;
  tn_value_reserved_1: SubCriterionScore;
  tn_value_reserved_2: SubCriterionScore;

  // Denormalized computed scores (1-5 weighted averages)
  complexity_score: number | null;
  value_creation_score: number | null;
  composite_score: number | null;

  // Derived from total_budget against admin t-shirt thresholds
  tshirt_size: TshirtSize | null;
  total_budget: number | null;

  // Active admin weights snapshot at read time
  weights: TechNavigatorWeights;
}

/**
 * Partial-update body for PUT /api/projects/{id}/tech-navigator.
 *
 * All fields optional. Sub-criteria can be set to null to clear a score.
 */
export interface TechNavigatorUpdate {
  project_type?: ProjectType | null;
  transformation_level?: TransformationLevel | null;
  tn_standardization?: SubCriterionScore;
  tn_usage?: SubCriterionScore;
  tn_maintenance?: SubCriterionScore;
  tn_financial_benefit?: SubCriterionScore;
  tn_payback?: SubCriterionScore;
  tn_competitive_advantage?: SubCriterionScore;
  tn_value_reserved_1?: SubCriterionScore;
  tn_value_reserved_2?: SubCriterionScore;
}
